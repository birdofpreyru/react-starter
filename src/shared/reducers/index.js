/**
 * This module holds the root factory for Redux reducer. The factory accepts
 * ExpressJS HTTP request object as its only optional argument, and it always
 * returns a promise that resolves to a new reducer.
 *
 * When the argument is provided, the factory assumes server-side rendering,
 * and it can use data available in the HTTP request to generate initial store
 * state fitted to the request.
 *
 * When no argument is provided, the factory assumes client-side rendering,
 * and creates the default state (in case of server-side rendering, that default
 * state will be overriden at client side by the one injected from the server).
 *
 * To better understand reducers read:
 * http://redux.js.org/docs/basics/Reducers.html.
 */

import { redux } from 'topcoder-react-utils';

/**
 * @param {Object} req Optional. ExpressJS HTTP request.
 * @return {Promise} Resolves to a new reducer.
 */
export async function factory(req) {
  const reducers = await redux.resolveReducers({
    /* The place to resolve child reducers created by factories. */
  });
  return redux.combineReducers((state) => {
    if (!req) return state;
    return {
      ...state,

      /* <MetaTags> rely on this field. */
      domain: `${req.protocol}://${req.headers.host || req.hostname}`,
    };
  }, {
    ...reducers,
  });
}

export default undefined;
