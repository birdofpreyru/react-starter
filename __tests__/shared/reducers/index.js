import { factory } from 'reducers';

describe('Reducer created without HTTP request', () => {
  let reducer;

  beforeAll(async () => {
    reducer = await factory();
  });

  test('Creates expected initial state', () => {
    expect(reducer(undefined, '@@INIT')).toMatchSnapshot();
  });
});


describe('Reducer created with HTTP request', () => {
  let reducer;

  beforeAll(async () => {
    reducer = await factory({
      headers: {

      },
      hostname: 'test-website.com',
      protocol: 'http',
    });
  });

  test('Creates expected initial state', () => {
    expect(reducer(undefined, '@@INIT')).toMatchSnapshot();
  });
});
