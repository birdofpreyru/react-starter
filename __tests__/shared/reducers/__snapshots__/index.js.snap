// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer created with HTTP request Creates expected initial state 1`] = `
Object {
  "domain": "http://test-website.com",
}
`;

exports[`Reducer created without HTTP request Creates expected initial state 1`] = `Object {}`;
